### ISSUE TITLE
Custom field component value not retained across renders

### ISSUE THREAD

**Issue by GeorgeyB** [payloadcms/payload#624]
Hi guys, I'm trying to swap out a field component, but the value doesn't seem to be retained between renders. Any ideas? Here's a component I'm using in its simplest form:

```
import React from 'react';
import { withCondition, useField } from 'payload/components/forms';
import { TextField } from 'payload/dist/fields/config/types';

function TestField({ path }: TextField & { path?: string }) {
    const { value, setValue } = useField<string>({
        path: path || ''
    });

    return (
        <input
            className="border py-1 px-4"
            type="text"
            value={value}
            onChange={({ target: { value } }) => setValue(value)}
        />
    );
}

export default withCondition(TestField);
```

As soon as a rerender occurs (in this case, my cursor leaves the parent array field) the value is lost.

**Comment by jmikrut** [comment]
Hey @GeorgeyB — this is strange. Do you have any other fields that may be impacting the data of this specific field's `path`? At first glance, it doesn't seem like there's anything wrong with your custom field. 

We will test around with custom fields within array fields, but we've got a lot of these in use across many projects and haven't seen this specific issue.

Let us report back shortly.

**Comment by GeorgeyB** [comment]
@jmikrut Good catch. The issue doesn't seem to occur when the component is for a top-level field - though even at the top level, the field seems to clear after an arbitary amount of time and then never again. Here's a screen recording to try and demonstrate:

https://user-images.githubusercontent.com/13889255/171512269-b884207f-2659-419c-bf7d-cfbe1dac5698.mp4

**Comment by GeorgeyB** [comment]
Thanks @DanRibbens!

I can confirm that this works as expected with [your fix](https://github.com/payloadcms/payload/pull/626).

**Comment by GeorgeyB** [comment]
Hi @DanRibbens,

this change seems to prevent any native React hooks from functioning correctly. Using this minimal example:

```
import React, { ChangeEventHandler, useCallback } from 'react';
import { useField } from 'payload/components/forms';

function TestField({ path, name }: any) {
    const { value, setValue } = useField<string>({
        path: path || name
    });

    const onChange: ChangeEventHandler<HTMLInputElement> = useCallback(
        ({ target: { value } }) => setValue(value),
        [setValue]
    );

    return (
        <input
            className="border py-1 px-4"
            type="text"
            value={value || ''}
            onChange={onChange}
        />
    );
}

export default TestField;
```

Throws a `dispatcher is null` error on the client:

![image](https://user-images.githubusercontent.com/13889255/172025610-b331d132-c4e3-4028-b27d-2fc7bbe1ddf1.png)

**Comment by DanRibbens** [comment]
Hey @GeorgeyB I can't reproduce this in `demo` using your exact component on a conditional field in a collection.

I suspect that your package.json has an issue with **1. You might have mismatching versions of React and the renderer**
We upgraded Payload to use React 18 recently so any other dependencies you have need to be on 18.

Would you mind sharing a repo that recreates this or a branch from Payload that shows it in `demo` for me to see?

**Comment by GeorgeyB** [comment]
Hi @DanRibbens, apologies for the delay coming back to you. React hooks are working as expected, but I'm still experiencing the initial issue whereby values are being lost between renders. I've been able to reproduce this in a completely fresh Payload install:

1. Initialise a new install with the `npx create-payload-app` command
2. Uncomment the "Examples" collection import and configuration item in `payload.config.ts`
3. Use the following config for the Examples collection (I've added the `testFields` array field and a `testField` sub field inside it):
```
import { CollectionConfig } from 'payload/types';

import CustomTextField from './CustomTextField';

// Example Collection - For reference only, this must be added to payload.config.ts to be used.
const Examples: CollectionConfig = {
    slug: 'examples',
    admin: {
        useAsTitle: 'someField'
    },
    fields: [
        {
            name: 'someField',
            type: 'text'
        },
        {
            name: 'testFields',
            type: 'array',
            fields: [
                {
                    name: 'testField',
                    type: 'text',
                    admin: {
                        components: {
                            Field: CustomTextField
                        }
                    }
                }
            ]
        }
    ]
};

export default Examples;
```

Where `CustomTextField` is essentially the example provided in the [Payload documentation](https://payloadcms.com/docs/admin/components#sending-and-receiving-values-from-the-form):

```
import { useField } from 'payload/components/forms';

const CustomTextField = ({ path }) => {
  const { value, setValue } = useField({ path });

  return (
    <input
      onChange={(e) => setValue(e.target.value)}
      value={value || ''}
    />
  )
}
```

The value shows in the input until the parent array field rerenders, then it is lost.

It's worth noting that even if `testField` is a top-level field, the value is still lost after some arbitary time/event, but only once (as shown in the video in [this comment](https://github.com/payloadcms/payload/issues/624#issuecomment-1144206643))

**Comment by DanRibbens** [comment]
Hey @GeorgeyB I haven't published the fix yet in the latest version yet. There were some other big changes merged at the same time and we wanted to do some additional QA on it first. I'll update again when its released.

**Comment by jmikrut** [comment]
@GeorgeyB — the fix is now deployed in `0.18.0`. 

Will you give it a shot, and if it resolves the problem you're facing, will you close this issue?

**Comment by GeorgeyB** [comment]
@jmikrut, @DanRibbens.

Looking good now, thanks guys.
